### PR TITLE
docstring: fix handling of empty documentation comments

### PR DESCRIPTION
**Pull request by jnikula**
Sometimes it's handy to add an empty documentation comment to generate documentation for a symbol without having to add some placeholder text. Or to include documentation for members/enumerations without documenting the struct/enum.

Currently, empty documentation comments lead to:

  File "docstring.py", line 143, in _remove_comment_markers
    while (not lines[0] or lines[0].isspace()):
               ~~~~~^^^
IndexError: list index out of range

Fix it and add some tests and documentation for it.

### CODE REVIEW

File: doc/syntax.rst [hunk 1]
```diff
@@ -46,6 +46,17 @@ includes macros, functions, struct and union members, enumerations, etc.
...
+----------------------------
+
+Use the empty documentation comment ``/***/`` to only generate the C or C++
+Domain directives without documentation. This can be useful for documenting
+struct members or enumeration constants without actually documenting the
+enclosing struct or enum.
+
+It's also a way to let Sphinx know about types to highlight them in
+e.g. parameter documentation.
+
```
Comment (BrunoMSantos) [inline doc/syntax.rst:59]: I wonder whether this is necessary at all, but I don't object to it.

File: test/c/comment-strip.c [hunk 2]
```diff
@@ -107,3 +107,27 @@ int two_liner(int foo, int bar);
...
+	/** Another empty documentation comment. */
+	int description;
+};
+
+/**
+ */
+struct empty_comment_3 {
+	/** Another empty documentation comment. */
+	int description;
+};
```
Comment (BrunoMSantos) [inline test/c/comment-strip.c:133]: What about top level comments? Any chance it might be an issue for `autosection`? (I plead ignorance by the way.)

**Comment by BrunoMSantos** [comment]
Interesting edge case. I think it's good over the covered cases and half expecting my worry about top level comments is unfounded. In which case, looks good to me.

Out of curiosity, is this just a pile of things you had laying around or is it being currently driven by actual use cases / requests?

**Comment by jnikula** [comment]
> Interesting edge case. I think it's good over the covered cases and half expecting my worry about top level comments is unfounded. In which case, looks good to me.
> 
> Out of curiosity, is this just a pile of things you had laying around or is it being currently driven by actual use cases / requests?

I was hacking on Mesa documentation a bit, see https://gitlab.freedesktop.org/mesa/mesa/-/merge_requests/23145#note_1985368

There were cases where Doxygen+Breathe happily generated documentation if struct members were documented but the struct itself was not. I'm not sure I like to promote that model in Hawkmoth, but I thought adding an empty documentation comment to the enclosing struct would be a nice workaround instead of having to add placeholder text (especially because I was unable to come up with *actual* documentation in the conversion). And it promptly crashed. :facepalm:

The warning improvements came from there as well. There were directives with typoed/wrong names or filenames, and there was no warnings, just silently no documentation.

File: doc/syntax.rst (hunk 1 above)
Comment (jnikula) [reply doc/syntax.rst:59]: I can leave that part out. I was contemplating leaving out the entire documentation on empty comments, but thought there's no harm.

File: test/c/comment-strip.c (hunk 2 above)
Comment (jnikula) [reply test/c/comment-strip.c:133]: Did not try yet. But seems like it could be a **feature**! You could use it to end a section (the future version with `:members:` support).

File: doc/syntax.rst (hunk 1 above)
Comment (BrunoMSantos) [reply doc/syntax.rst:59]: I meant the whole section too by the way. I see no point in documenting this at all, I think anyone would expect this behaviour intuitively.

File: test/c/comment-strip.c (hunk 2 above)
Comment (BrunoMSantos) [reply test/c/comment-strip.c:133]: Ahahah, maybe...

I'm expecting it to work fine right now, but I'd welcome a unit test for selecting sections before and after one of these at least. In case we inadvertently break it later on.

File: doc/syntax.rst (hunk 1 above)
Comment (jnikula) [reply doc/syntax.rst:59]: Maybe just one sentence in the previous section?

"Empty documentation comments `/***/` create the C or C++ Domain directives without documentation."

Or nothing?

Comment (BrunoMSantos) [reply doc/syntax.rst:59]: Up to you... I don't feel it's needed but I'd prefer a single sentence to a whole section.

File: test/c/comment-strip.c (hunk 2 above)
Comment (jnikula) [reply test/c/comment-strip.c:137]: Dropped it completely. Can be added later if needed.

File: doc/syntax.rst (hunk 1 above)
Comment (jnikula) [reply doc/syntax.rst]: Dropped it completely. Can be added later if needed.

File: test/c/comment-strip.c (hunk 2 above)
Comment (BrunoMSantos) [reply test/c/comment-strip.c:137]: Yeah, feels complete now. Thanks.
